File: www/cgi/hello.py
#!/usr/bin/env python3

import os
import sys
from urllib.parse import parse_qs

# Print the Content-Type header
print("Content-Type: text/html\n")

# HTML structure
print("<html><head><title>CGI Hello</title></head><body>")
print("<h1>Hello, World! This is Python!</h1>")

# Debug: Print environment variables (remove in production)
print("<h2>Environment Variables</h2>")
print("<pre>")
for key, value in os.environ.items():
    print(f"{key}={value}")
print("</pre>")

# Check the request method
if os.environ.get("REQUEST_METHOD") == "POST":
    # Read POST data
    content_length = int(os.environ.get("CONTENT_LENGTH", 0))
    post_data = sys.stdin.read(content_length)
    
    # Debug: Print POST data
    print("<h2>Debug POST Data</h2>")
    print(f"<pre>{post_data}</pre>")
    
    # Parse POST data
    parsed_data = parse_qs(post_data)
    name = parsed_data.get("name", [""])[0]
    message = parsed_data.get("message", [""])[0]
    
    # Display parsed POST data
    print("<h2>Received POST Data</h2>")
    print(f"<p>Name: {name}</p>")
    print(f"<p>Message: {message}</p>")
else:
    print("<p>No POST data received.</p>")

# Close HTML structure
print("</body></html>")

File: www/cgi/hello.php
#!/usr/bin/env php
<?php
// 1) Ensure we have argv[1]:
if ($argc < 2 || empty($argv[1])) {
    echo "Its rude not to introduce yourself! Please tell me your name.\n";
    exit(0);
}

// 2) The full raw HTTP request is in argv[1].
$input = $argv[1];

// 3) Separate headers from body by "\r\n\r\n"
$parts = explode("\r\n\r\n", $input, 2);

// If we have two parts, the second is the body; otherwise use the entire input as the body.
if (count($parts) === 2) {
    list($rawHeaders, $rawBody) = $parts;
} else {
    // If there's no blank line, everything is 'body'
    $rawBody = $input;
}

// 4) Output minimal CGI headers
header("Content-Type: text/html");
echo "\r\n";  // End of headers
if (empty($rawBody)){

    echo "Its rude not to introduce yourself! Please tell me your name.\n";
}
else {
    echo "Hello " . htmlspecialchars($rawBody) . "!\r\n";
}

// 5) Display the body in a "Hello ...!" message
// echo "<html><head><title>Hello CGI</title></head><body>";
// echo "</body></html>";
